beacon-chain/db/kv/regen_historical_states.go

```diff
@@ -0,0 +1,180 @@
+package kv
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	ethpb "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
+	"github.com/prysmaticlabs/go-ssz"
+	"github.com/prysmaticlabs/prysm/beacon-chain/core/helpers"
+	transition "github.com/prysmaticlabs/prysm/beacon-chain/core/state"
+	"github.com/prysmaticlabs/prysm/beacon-chain/db/filters"
+	stateTrie "github.com/prysmaticlabs/prysm/beacon-chain/state"
+	"github.com/prysmaticlabs/prysm/shared/params"
+	log "github.com/sirupsen/logrus"
+	"go.opencensus.io/trace"
+)
+
+func (kv *Store) regenHistoricalStates(ctx context.Context) error {
+	ctx, span := trace.StartSpan(ctx, "db.regenHistoricalStates")
+	defer span.End()
+
+	b, err := kv.HighestSlotBlocks(ctx)
+	if err != nil {
+		return err
+	}
+	if len(b) == 0 {
+		return errors.New("blocks can't be empty")
+	}
+	if b[0] == nil {
+		return errors.New("nil last block")
+	}
+	lastSavedBlockSlot := b[0].Block.Slot
+	slotsPerArchivedPoint := params.BeaconConfig().SlotsPerArchivedPoint
+	lastSavedBlockArchivedIndex := lastSavedBlockSlot/slotsPerArchivedPoint - 1
+
+	// Start from genesis point.
+	genesisState, err := kv.GenesisState(ctx)
+	if err != nil {
+		return err
+	}
+	if genesisState == nil {
+		return errors.New("nil genesis state")
+	}
+	currentState := genesisState.Copy()
+	startSlot := genesisState.Slot()
+	lastArchivedIndex, err := kv.LastArchivedIndex(ctx)
+	if err != nil {
+		return err
+	}
+
+	// Restore from last archived point if this process was previously interrupted.
+	if lastArchivedIndex > 0 {
+		states, err := kv.HighestSlotStatesBelow(ctx, (lastArchivedIndex-1)*slotsPerArchivedPoint+1)
```

Comment:
Can we make this second argument into its own variable to understand why we do the -1 and the * +1?

Reply:
Yep. Done